pkg/schedule/schedulers/balance_range.go

```diff
@@ -167,6 +230,286 @@ func newBalanceRangeScheduler(opController *operator.Controller, conf *balanceRa
+	opInfluence  operator.OpInfluence
+	tolerate     int64
+}
+
+func (s *balanceRangeScheduler) prepare(cluster sche.SchedulerCluster, opInfluence operator.OpInfluence, job *balanceRangeSchedulerJob) (*balanceRangeSchedulerPlan, error) {
+	filters := s.filters
+	switch job.Engine {
+	case core.EngineTiKV:
+		filters = append(filters, filter.NewEngineFilter(string(types.BalanceRangeScheduler), filter.NotSpecialEngines))
+	case core.EngineTiFlash:
+		filters = append(filters, filter.NewEngineFilter(string(types.BalanceRangeScheduler), filter.SpecialEngines))
+	default:
+		return nil, errs.ErrGetSourceStore.FastGenByArgs(job.Engine)
+	}
+	sources := filter.SelectSourceStores(cluster.GetStores(), filters, cluster.GetSchedulerConfig(), nil, nil)
+	if sources == nil {
+		return nil, errs.ErrStoresNotEnough.FastGenByArgs("no store to select")
+	}
+
+	krs := core.NewKeyRanges(job.Ranges)
+	scanRegions, err := cluster.BatchScanRegions(krs)
+	if err != nil {
+		return nil, err
+	}
+
+	// storeID <--> score mapping
+	scoreMap := make(map[uint64]int64, len(sources))
+	for _, source := range sources {
+		scoreMap[source.GetID()] = 0
+	}
+	totalScore := int64(0)
+	for _, region := range scanRegions {
+		for _, peer := range job.Role.getPeers(region) {
+			scoreMap[peer.GetStoreId()] += 1
+			totalScore += 1
+		}
+	}
+
+	sort.Slice(sources, func(i, j int) bool {
+		role := job.Role
+		iop := role.getStoreInfluence(opInfluence.GetStoreInfluence(sources[i].GetID()))
+		jop := role.getStoreInfluence(opInfluence.GetStoreInfluence(sources[j].GetID()))
+		iScore := scoreMap[sources[i].GetID()]
+		jScore := scoreMap[sources[j].GetID()]
+		return iScore+iop > jScore+jop
+	})
+
+	averageScore := int64(0)
+	if len(sources) != 0 {
+		averageScore = totalScore / int64(len(sources))
+	}
+
+	tolerate := int64(float64(len(scanRegions)) * adjustRatio)
+	if tolerate < 1 {
+		tolerate = 1
+	}
+	return &balanceRangeSchedulerPlan{
+		SchedulerCluster: cluster,
+		stores:           sources,
+		scoreMap:         scoreMap,
+		source:           nil,
+		target:           nil,
+		region:           nil,
+		averageScore:     averageScore,
+		job:              job,
+		opInfluence:      opInfluence,
+		tolerate:         tolerate,
+	}, nil
+}
+
+func (p *balanceRangeSchedulerPlan) sourceStoreID() uint64 {
+	return p.source.GetID()
+}
+
+func (p *balanceRangeSchedulerPlan) targetStoreID() uint64 {
+	return p.target.GetID()
+}
+
+func (p *balanceRangeSchedulerPlan) score(storeID uint64) int64 {
+	return p.scoreMap[storeID]
+}
+
+func (p *balanceRangeSchedulerPlan) shouldBalance(scheduler string) bool {
+	sourceInfluence := p.opInfluence.GetStoreInfluence(p.sourceStoreID())
+	sourceInf := p.job.Role.getStoreInfluence(sourceInfluence)
+	if sourceInf < 0 {
+		sourceInf = -sourceInf
+	}
+	sourceScore := p.sourceScore - sourceInf - p.tolerate
+
+	targetInfluence := p.opInfluence.GetStoreInfluence(p.targetStoreID())
+	targetInf := p.job.Role.getStoreInfluence(targetInfluence)
+	if targetInf < 0 {
```

Comment:
targetInf is int64?

Reply:
yes, if the store has some remove-peer operators

Reply:
So we need not check `targetInf < 0`?

Reply:
I must consider the scenario involving `removing-peer `, so the `targeting` is negative, the target score is small. This causes too many peers to be added to this store.